### PR TITLE
Backport securerandom

### PR DESCRIPTION
**Pull request by headius**
This PR backports recent SecureRandom entropy fixes from 9k to 1.7. See #1405.

We had a 1.7 user in #jruby today with entropy problems.

### CODE REVIEW

**Comment by kares** [comment]
isn't https://github.com/jruby/jruby/pull/4149/commits/8fb4b9314dff380b6bccd5174bf456d0b6c271d7 "dangerous" to back-port (changing a previously initialized field on `ThreadContext` to null) ?

File: core/src/main/java/org/jruby/util/cli/Options.java [hunk 1]
```diff
@@ -146,6 +146,7 @@
     public static final Option<Boolean> REIFY_VARIABLES = bool(MISCELLANEOUS, "reify.variables", false, "Attempt to expand instance vars into Java fields");
     public static final Option<Boolean> PREFER_IPV4 = bool(MISCELLANEOUS, "net.preferIPv4", true, "Prefer IPv4 network stack");
     public static final Option<Boolean> FCNTL_LOCKING = bool(MISCELLANEOUS, "file.flock.fcntl", true, "Use fcntl rather than flock for File#flock");
+    public static final Option<String> PREFERRED_PRNG = string(MISCELLANEOUS, "preferred.prng", "NativePRNGNonBlocking", "Maintain children static scopes to support scope dumping.");
```
Comment (kares) [inline core/src/main/java/org/jruby/util/cli/Options.java]: the help message for the option seems like a copy :spaghetti: left over

Comment (headius) [inline core/src/main/java/org/jruby/util/cli/Options.java]: Indeed...I'll fix that on both master and in this PR.

**Comment by headius** [comment]
> @kares: isn't 8fb4b93 "dangerous" to back-port (changing a previously initialized field on ThreadContext to null) ?

A little dangerous, I agree. However the internals of ThreadContext have never been a blessed API. I know there might be people poking at them, but we've never shown people examples that use TC internals and we have generally discouraged messing with it.

I'm not too concerned, especially since it helps thread startup times (important for Fiber and Enumerator#next as well). @enebo?

**Comment by headius** [comment]
@takanuva This is the backported SecureRandom changes I mentioned. Since you have an environment with entropy problems, perhaps you could test out a HEAD build of JRuby 1.7 for us once we merge?

It would be REALLY nice to know that this actually fixes your issue without installing an entropy-generator. Up to now we've just assumed it does, since it's hard to simulate an environment with entropy-starvation problems.

**Comment by enebo** [comment]
@bbrowning or @mkristian can you see a problem with this change to threadcontext not having securerandom initialized?  I personally think the potential for this breaking stuff could only happen in a deep extension but not very likely.

**Comment by bbrowning** [comment]
@enebo I don't know of anything that uses securerandom directly like this. I agree it would have to be an extension really digging deep to get hit by this. While it is possible that one exists somewhere, a quick GitHub search didn't turn up anything obvious.

**Comment by kares** [comment]
... no (more) worries from me than - thanks everyone for looking into it!

**Comment by takanuva** [comment]
Sure, @headius, just give me a few days and I'll gladly test that for you guys. 😄

**Pull request merged**